rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,372 @@
+    at implementation-defined positions.
+
+#### Constraints
+
+* (C?) `same(shape(inputs...))`.
+* (C?) `element_type(inputs...) = element_type(init_values...)`.
+* (C?) `baseline_element_type(inputs...) = baseline_element_type(results...)`.
+* (C?) `body` has type `tensor<E0>, ..., tensor<EN-1>, tensor<E0>, ...,`
+       `tensor<EN-1>) -> (tensor<E0>, ..., tensor<EN-1>)` where
+       `is_integer(element_type(inputs[i])) = is_integer(element_type(Ei]` or
+       `is_float(element_type(inputs[i])) = is_float(element_type(Ei]` or
+       `is_complex(element_type(inputs[i])) = is_complex(element_type(Ei]` or
+       `is_quantized(element_type(inputs[i])) = is_quantized(element_type(Ei]`.
+* (C?) `shape(results...) = shape(inputs...)` except that the dimension
+  sizes of `inputs...` corresponding to `dimensions` are not included.
+
+`reduce_implicit_convert` is defined as
+
+```python
+def reduce_implicit_convert(x: Value, source_type: Type, destination_type:
+        Type):
+    if source_type == destination_type:
+        return x
+    if is_quantized(source_type) and is_quantized(destination_type):
+        return quantize(x, destination_type)
+    return convert(x, destination_type)
+```
+
+The above specification of `reduce` op can be used to define the specification
+of other ops as shown below. As before, we are only presenting the relevant
+portions of the spec which needs modification.
+
+### Revised specification of  reduce_window op
+
+#### Constraints
+
+* (C?) `element_type(inputs...) = element_type(init_values...)`.
+* (C?) `baseline_element_type(inputs...) = baseline_element_type(results...)`.
+* (C?) `body` has type `tensor<E0>, ..., tensor<EN-1>, tensor<E0>, ...,`
+       `tensor<EN-1>) -> (tensor<E0>, ..., tensor<EN-1>)` where
+       `is_integer(element_type(inputs[i])) = is_integer(element_type(Ei]` or
+       `is_float(element_type(inputs[i])) = is_float(element_type(Ei]` or
+       `is_complex(element_type(inputs[i])) = is_complex(element_type(Ei]` or
+       `is_quantized(element_type(inputs[i])) = is_quantized(element_type(Ei]`.
+
+### Revised specification of select_and_scatter op
+
+This op originally takes two function arguments `select` and `scatter`. As the
+`select` function is supposed to perform a non-accumulative operation, we may
+not need additional conversion functions associated with `select`. But the
+`scatter` function needs be accompanied with `input_conversion` and
+`output_conversion` functions.
+
+#### Constraints
+
+<!-- markdownlint-disable line-length -->
+* (C1) `element_type(operand) = element_type(source)`.
+* (C3) `element_type(init_value) = element_type(operand)`.
+* (C?) `baseline_element_type(inputs...) = baseline_element_type(results...)`.
+* (C10) `scatter` has type `(tensor<E>, tensor<E>) -> tensor<E>` where
+       `is_integer(element_type(operand)) = is_integer(element_type(E]` or
```

Comment:
s/element_type(E/element_type(E)/

(note: missing `)`)

and below

Reply:
done.